### PR TITLE
Adding AmbrosM's solution to methods

### PR DESCRIPTION
**Pull request by andrew-benz**

### CODE REVIEW

File: scripts/.ipynb_checkpoints/download_resources-checkpoint.sh
Comment (rcannood) [inline scripts/.ipynb_checkpoints/download_resources-checkpoint.sh:1]: Could you remove the various `.ipynb_checkpoints` directories created in this PR? :relaxed:

File: src/task/methods/ambros_m_judges_third_place/config.vsh.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
+__merge__: ../../api/comp_method.yaml
+
+functionality:
+  name: ambros_m_judges_third_place
+  info:
+    label: An ensemble model developed by AmbrosM
+    summary: "Ensemble comprised of four models: a py-boost model, an ExtraTrees model, and two recommender system models."
```
Comment (rcannood) [inline src/task/methods/ambros_m_judges_third_place/config.vsh.yaml:7]: I propose adding a longer description

```suggestion
    description: |
      An ensemble of four models for single-cell perturbation prediction: 
      Py-boost (a ridge regression-based recommender system),
      ExtraTrees (a decision tree ensemble with target-encoded features),
      a k-nearest neighbors recommender system, and a ridge regression model. 
      Each model offered distinct strengths and weaknesses: ExtraTrees and
      knn were unable to extrapolate beyond the training data, while ridge
      regression provided extrapolation capability. To enhance model performance,
      data augmentation techniques were used, including averaging differential
      expressions for compound mixtures and adjusting cell counts to reduce biases.
      The final ensemble combined predictions from all models, achieving competitive
      scores in cross-validation and the leaderboard.
```

File: src/task/methods/ambros_m_judges_third_place/config.vsh.yaml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+#     default: "resources/neurips-2023-kaggle/de_train.parquet"
+#   - type: string
+#     name: --train_obs_zip
+#     default: "resources/neurips-2023-kaggle/train_obs.csv.zip"
+#   - type: string
+#     name: --id_map
+#     default: "resources/neurips-2023-kaggle/id_map.csv"
+#   - type: string
+#     name: --output
+#     default: "output.parquet"
```
Comment (rcannood) [inline src/task/methods/ambros_m_judges_third_place/config.vsh.yaml:22]: This section can be removed

```suggestion
```

File: src/task/methods/ambros_m_judges_third_place/script.py [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+pd.set_option("min_rows", 6)
+
+## Viash start
+
+par = dict(
+    de_train_path = "/home/jovyan/git-repos/task-dge-perturbation-prediction/resources/neurips-2023-kaggle/de_train.parquet",
+    train_obs_zip_path = "/home/jovyan/git-repos/task-dge-perturbation-prediction/resources/neurips-2023-kaggle/train_obs.csv.zip",
+    id_map_path = "/home/jovyan/git-repos/task-dge-perturbation-prediction/resources/neurips-2023-kaggle/id_map.csv",
+    output_path = "output.parquet",
+)
```
Comment (rcannood) [inline src/task/methods/ambros_m_judges_third_place/script.py:35]: Turn this into a proper Viash placeholder. Viash will detect the `## VIASH START` -- `## VIASH END` tags and replace it with an argument parser when the component is transformed into an executable or a nextflow module.

```suggestion
## VIASH START
par = dict(
    de_train_h5ad = "resources/neurips-2023-data/de_train.h5ad",
    id_map = "resources/neurips-2023-data/id_map.csv",
    output_path = "output.parquet",
)
## VIASH END
```

File: src/task/methods/ambros_m_judges_third_place/script.py [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+    """
+#     p_value = scipy.stats.t.cdf(- np.abs(t_score), df=420) * 2
+    p_value = norm.cdf(- np.abs(t_score)) * 2
+    p_value = p_value.clip(1e-180, None)
+    return - np.log10(p_value) * np.sign(t_score)
+
+## Loading data
+
+de_train = pd.read_parquet(par['de_train_path'])
+id_map = pd.read_csv(par['id_map_path'], index_col = 0)
```
Comment (rcannood) [inline src/task/methods/ambros_m_judges_third_place/script.py:78]: Rename args in line with what is defined in `../api/comp_method.yaml`.

```suggestion
de_train_h5ad = pd.read_parquet(par['de_train_h5ad'])
id_map = pd.read_csv(par['id_map'], index_col = 0)
```

File: src/task/methods/ambros_m_judges_third_place/config.vsh.yaml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  resources:
+    - type: python_script
+      path: script.py
+platforms:
+  - type: docker
+    image: ghcr.io/openproblems-bio/base_python:1.0.4
```
Comment (rcannood) [inline src/task/methods/ambros_m_judges_third_place/config.vsh.yaml:28]: This image already has cuda & pytorch installed

```suggestion
    image: ghcr.io/openproblems-bio/base_pytorch_nvidia:1.0.4
```

File: src/task/methods/ambros_m_judges_third_place/config.vsh.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    setup:
+      - type: python
+        packages:
+          - anndata==0.10.3
```
Comment (rcannood) [inline src/task/methods/ambros_m_judges_third_place/config.vsh.yaml:32]: The openproblems base image already contains a specific version of anndata

```suggestion
```

File: src/task/methods/ambros_m_judges_third_place/config.vsh.yaml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+          - array-api-compat==1.4
+          - async-generator==1.10
+          - cupy-cuda11x==12.2.0
```
Comment (rcannood) [inline src/task/methods/ambros_m_judges_third_place/config.vsh.yaml:35]: base_pytorch_nvidia contains cuda 12.x, not 11.x, so I need to update this, too

```suggestion
          - cupy-cuda12x
```

File: src/task/methods/ambros_m_judges_third_place/config.vsh.yaml [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      path: script.py
+platforms:
+  - type: docker
+    image: ghcr.io/openproblems-bio/base_python:1.0.4
+    setup:
+      - type: python
+        packages:
+          - anndata==0.10.3
+          - array-api-compat==1.4
+          - async-generator==1.10
```
Comment (rcannood) [inline src/task/methods/ambros_m_judges_third_place/config.vsh.yaml:34]: When I try to build it complains about colorama not being installed

```suggestion
          - async-generator==1.10
          - colorama
```

File: src/task/methods/ambros_m_judges_third_place/script.py [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+import scipy
+from sklearn.linear_model import Ridge
+from sklearn.decomposition import PCA
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import StandardScaler
+from sklearn.neighbors import KNeighborsRegressor
+from sklearn.ensemble import ExtraTreesRegressor
+from itertools import combinations
+import cupy as cp
+from py_boost import GradientBoosting
```
Comment (rcannood) [inline src/task/methods/ambros_m_judges_third_place/script.py:20]: Is there an alternative to GradientBoosting we can use when there is no GPU available?

**Review by rcannood — CHANGES_REQUESTED** [review]
Already a partial review; since I'm still working on other things I might already send you what I have so far :)

File: src/task/methods/ambros_m_judges_third_place/script.py (hunk 4 above)
Comment (andrew-benz) [reply src/task/methods/ambros_m_judges_third_place/script.py:77]: In this case I should use `de_train = pd.read_parquet(par['de_train'])`, right? The method uses the parquet version of `de_train`, not the .h5ad version

**Comment by andrew-benz** [comment]
@rcannood Unfortunately the solution heavily depends on `py_boost` (https://github.com/sberbank-ai-lab/Py-Boost), which I was only able to run with CUDA 11 (not CUDA 12) and which requires a GPU. As far as I know there is no library with the exact same functionality but with CPU support that we could easily incorporate into this solution.

**Comment by andrew-benz** [comment]
Looks like the `test / viash test` check failed due to the `llvm-lite` install:
```#5 1.887 Collecting llvmlite==0.39.1
#5 1.908   Downloading llvmlite-0.39.1.tar.gz (132 kB)
#5 1.946      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.8/132.8 kB 4.1 MB/s eta 0:00:00
#5 1.969   Preparing metadata (setup.py): started
#5 2.183   Preparing metadata (setup.py): finished with status 'error'
#5 2.186   error: subprocess-exited-with-error
#5 2.186   
#5 2.186   × python setup.py egg_info did not run successfully.
#5 2.186   │ exit code: 1
#5 2.186   ╰─> [8 lines of output]
#5 2.186       Traceback (most recent call last):
#5 2.186         File "<string>", line 2, in <module>
#5 2.186         File "<pip-setuptools-caller>", line 34, in <module>
#5 2.186         File "/tmp/pip-install-vx3d4fie/llvmlite_0097ae8b9f1e4097864f0886e2d76a65/setup.py", line 55, in <module>
#5 2.186           _guard_py_ver()
#5 2.186         File "/tmp/pip-install-vx3d4fie/llvmlite_0097ae8b9f1e4097864f0886e2d76a65/setup.py", line 52, in _guard_py_ver
#5 2.186           raise RuntimeError(msg.format(cur_py, min_py, max_py))
#5 2.186       RuntimeError: Cannot install on Python version 3.11.9; only versions >=3.7,<3.11 are supported.
#5 2.186       [end of output]
#5 2.186   
#5 2.186   note: This error originates from a subprocess, and is likely not a problem with pip.
#5 2.188 error: metadata-generation-failed
```

Is there any way to adjust the version of Python for just this component?

**Review by rcannood — APPROVED** [review]
LGTM

File: src/task/methods/pyboost/script.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+## VIASH START
+par = dict(
+    de_train = "resources/neurips-2023-data/de_train.parquet",
+    id_map = "resources/neurips-2023-data/id_map.csv",
+    train_obs_zip = "resources/neurips-2023-kaggle/train_obs.csv.zip",
+    predictor_names = ["py_boost"],
+    output = "output.parquet",
+)
+meta = dict(
+    resources_dir = "src/task/methods/ambros_m_judges_third_place"
```
Comment (andrew-benz) [inline src/task/methods/pyboost/script.py:23]: Should this be changed to `src/task/methods/pyboost`?

Comment (rcannood) [reply src/task/methods/pyboost/script.py]: Fixed!

**Pull request merged**